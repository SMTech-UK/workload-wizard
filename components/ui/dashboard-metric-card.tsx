import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-500">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-muted-foreground mt-1">{subtitle}</div>}
    </div>
  );
} 