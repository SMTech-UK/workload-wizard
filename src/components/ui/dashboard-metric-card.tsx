import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white dark:bg-zinc-900 shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-900 dark:text-white">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-gray-600 dark:text-gray-300 mt-1">{subtitle}</div>}
    </div>
  );
} 