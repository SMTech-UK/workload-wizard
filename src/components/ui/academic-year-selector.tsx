"use client"

import React from "react";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
import { Badge } from "@/components/ui/badge";
import { useAcademicYear } from "@/hooks/useAcademicYear";
import { Calendar } from "lucide-react";

interface AcademicYearSelectorProps {
  className?: string;
  showActiveBadge?: boolean;
  onAcademicYearChange?: (academicYearId: string) => void;
}

export function AcademicYearSelector({ 
  className, 
  showActiveBadge = true,
  onAcademicYearChange 
}: AcademicYearSelectorProps) {
  const { 
    academicYears, 
    currentAcademicYear, 
    currentAcademicYearId, 
    changeAcademicYear 
  } = useAcademicYear();

  const handleAcademicYearChange = async (academicYearId: string) => {
    try {
      await changeAcademicYear(academicYearId);
      onAcademicYearChange?.(academicYearId);
    } catch (error) {
      console.error('Failed to change academic year:', error);
    }
  };

  // Function to format academic year name (e.g., "2025/26" -> "AY-25/26")
  const formatAcademicYearName = (yearName: string) => {
    // Extract the last 2 digits from each year part
    const parts = yearName.split('/');
    if (parts.length === 2) {
      const startYear = parts[0].slice(-2);
      const endYear = parts[1].slice(-2);
      return `AY-${startYear}/${endYear}`;
    }
    // Fallback to original name if format is unexpected
    return yearName;
  };

  if (!academicYears || academicYears.length === 0) {
    return (
      <div className={`flex items-center gap-2 text-muted-foreground ${className}`}>
        <Calendar className="h-4 w-4" />
        <span className="text-sm">No academic years configured</span>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      <Select
        value={currentAcademicYearId || ""}
        onValueChange={handleAcademicYearChange}
      >
        <SelectTrigger className="w-auto min-w-[120px] flex items-center gap-2">
          <Calendar className="h-4 w-4 text-muted-foreground" />
          <span className="font-medium">
            {currentAcademicYear ? formatAcademicYearName(currentAcademicYear.name) : "Select AY"}
          </span>
        </SelectTrigger>
        <SelectContent>
          {academicYears.map((year: any) => (
            <SelectItem key={year._id} value={year._id}>
              <div className="flex items-center gap-2">
                <Calendar className="h-4 w-4 text-muted-foreground" />
                <span>{formatAcademicYearName(year.name)}</span>
                {year.isActive && showActiveBadge && (
                  <Badge variant="default" className="text-xs">Active</Badge>
                )}
                {year.isStaging && (
                  <Badge variant="secondary" className="text-xs">Staging</Badge>
                )}
              </div>
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
    </div>
  );
} 