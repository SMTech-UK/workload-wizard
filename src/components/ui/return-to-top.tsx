"use client";

import { Button } from "@/components/ui/button";
import { ChevronUp, Home } from "lucide-react";
import { useState, useEffect } from "react";
import Link from "next/link";

export function ReturnToTop() {
  const [isVisible, setIsVisible] = useState(true); // Temporarily always visible for testing

  useEffect(() => {
    const toggleVisibility = () => {
      console.log("Scroll position:", window.pageYOffset);
      if (window.pageYOffset > 300) {
        console.log("Setting visible to true");
        setIsVisible(true);
      } else {
        console.log("Setting visible to false");
        setIsVisible(false);
      }
    };

    // Call once on mount to check initial state
    toggleVisibility();
    
    window.addEventListener("scroll", toggleVisibility);
    return () => window.removeEventListener("scroll", toggleVisibility);
  }, []);

  const scrollToTop = () => {
    console.log("Scrolling to top");
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  console.log("ReturnToTop component render, isVisible:", isVisible);

  return (
    <>
      {/* Return to Home button - top left */}
      <Link href="/">
        <Button
          size="sm"
          variant="outline"
          className="fixed top-4 left-4 z-50 shadow-lg hover:shadow-xl transition-all duration-200"
          aria-label="Return to home"
        >
          <Home className="h-4 w-4 mr-2" />
          Home
        </Button>
      </Link>

      {/* Return to Top button - bottom left */}
      <Button
        onClick={scrollToTop}
        size="sm"
        className="fixed bottom-4 left-4 z-50 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 bg-blue-500 text-white"
        aria-label="Return to top"
      >
        <ChevronUp className="h-4 w-4" />
      </Button>
    </>
  );
} 