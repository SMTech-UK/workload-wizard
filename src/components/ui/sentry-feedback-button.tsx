"use client";

import { useEffect, useState } from "react";
import * as Sentry from "@sentry/nextjs";
import { Button } from "@/components/ui/button";
import { Bug, MessageCircle } from "lucide-react";
import { useModalAwareButton } from "@/hooks/useModalAwareButton";
import { createPortal } from "react-dom";

export function SentryFeedbackButton() {
  const [feedback, setFeedback] = useState<any>(null);
  const [widget, setWidget] = useState<any>(null);
  const [mounted, setMounted] = useState(false);
  const buttonRef = useModalAwareButton();

  useEffect(() => {
    setMounted(true);
    // Get the feedback instance on the client side
    setFeedback(Sentry.getFeedback());
  }, []);

  const handleFeedbackClick = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    if (!feedback) return;

    if (widget) {
      // If widget is open, close it
      widget.removeFromDom();
      setWidget(null);
    } else {
      // Create and show the form immediately
      feedback.createForm().then((newWidget: any) => {
        newWidget.appendToDom();
        newWidget.open();
        setWidget(newWidget);
      }).catch((error: any) => {
        console.error('Failed to create Sentry feedback form:', error);
      });
    }
  };

  if (!feedback || !mounted) return null;

  const button = (
    <Button
      ref={buttonRef}
      variant="destructive"
      size="lg"
      onClick={handleFeedbackClick}
      onMouseDown={(e) => {
        e.preventDefault();
        e.stopPropagation();
      }}
      className="fixed bottom-4 right-4 z-[99999] shadow-lg"
      style={{
        zIndex: 99999,
        pointerEvents: 'auto',
      }}
      aria-label="Report a bug or give feedback"
    >
      <Bug className="h-4 w-4 mr-2" />
      Report an Issue
    </Button>
  );

  // Render the button in a portal to ensure it's outside modal context
  return createPortal(button, document.body);
} 