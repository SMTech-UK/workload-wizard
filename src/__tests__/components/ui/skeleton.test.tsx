import { render, screen } from '@testing-library/react';
import { Skeleton } from '@/components/ui/skeleton';

// Custom query to find elements by data-slot attribute
const getByDataSlot = (container: HTMLElement, slotName: string) => {
  return container.querySelector(`[data-slot="${slotName}"]`);
};

const getAllByDataSlot = (container: HTMLElement, slotName: string) => {
  return container.querySelectorAll(`[data-slot="${slotName}"]`);
};

describe('Skeleton Component - Comprehensive UI Testing', () => {
  describe('Basic Rendering and Functionality', () => {
    it('should render skeleton with default styling', () => {
      const { container } = render(<Skeleton />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });

    it('should render skeleton with custom className', () => {
      const { container } = render(<Skeleton className="custom-skeleton" />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveClass('custom-skeleton');
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });

    it('should render skeleton with children', () => {
      const { container } = render(<Skeleton>Loading content</Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveTextContent('Loading content');
    });

    it('should render skeleton with custom dimensions', () => {
      const { container } = render(<Skeleton className="w-32 h-8" />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveClass('w-32', 'h-8');
    });
  });

  describe('Accessibility Testing', () => {
    it('should have proper data attribute', () => {
      const { container } = render(<Skeleton />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveAttribute('data-slot', 'skeleton');
    });

    it('should support ARIA attributes', () => {
      render(<Skeleton aria-label="Loading indicator" />);

      const skeleton = screen.getByLabelText('Loading indicator');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveAttribute('data-slot', 'skeleton');
    });

    it('should support role attribute', () => {
      render(<Skeleton role="status" aria-label="Loading" />);

      const skeleton = screen.getByRole('status');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveAttribute('data-slot', 'skeleton');
    });

    it('should have proper semantic structure', () => {
      const { container } = render(<Skeleton />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton?.tagName).toBe('DIV');
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });
  });

  describe('Responsive Behavior', () => {
    it('should maintain styling on mobile screens', () => {
      // Mock mobile screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375,
      });

      const { container } = render(<Skeleton className="w-full h-4" />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveClass('w-full', 'h-4');
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });

    it('should maintain styling on tablet screens', () => {
      // Mock tablet screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 768,
      });

      const { container } = render(<Skeleton className="w-64 h-6" />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveClass('w-64', 'h-6');
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });

    it('should maintain styling on desktop screens', () => {
      // Mock desktop screen size
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1024,
      });

      const { container } = render(<Skeleton className="w-96 h-8" />);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveClass('w-96', 'h-8');
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });
  });

  describe('Content Structure', () => {
    it('should handle text content', () => {
      const { container } = render(<Skeleton>Loading text content</Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveTextContent('Loading text content');
    });

    it('should handle HTML content', () => {
      const { container } = render(<Skeleton><span>Loading span</span></Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toContainHTML('<span>Loading span</span>');
    });

    it('should handle complex nested content', () => {
      const { container } = render(
        <Skeleton>
          <div>
            <h3>Loading Title</h3>
            <p>Loading description</p>
          </div>
        </Skeleton>
      );

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveTextContent('Loading Title');
      expect(skeleton).toHaveTextContent('Loading description');
    });

    it('should handle empty content', () => {
      const { container } = render(<Skeleton></Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toBeInTheDocument();
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });
  });

  describe('Edge Cases and Error Handling', () => {
    it('should handle very long content', () => {
      const longContent = 'A'.repeat(500);

      const { container } = render(<Skeleton>{longContent}</Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveTextContent(longContent);
      expect(skeleton).toHaveClass('bg-accent', 'animate-pulse', 'rounded-md');
    });

    it('should handle special characters in content', () => {
      const specialContent = 'Loading with special chars: <>&"\'<>';

      const { container } = render(<Skeleton>{specialContent}</Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveTextContent(specialContent);
    });

    it('should handle multiple skeletons', () => {
      const { container } = render(
        <div>
          <Skeleton>First skeleton</Skeleton>
          <Skeleton>Second skeleton</Skeleton>
          <Skeleton>Third skeleton</Skeleton>
        </div>
      );

      const skeletons = getAllByDataSlot(container, 'skeleton');
      expect(skeletons).toHaveLength(3);
      expect(skeletons[0]).toHaveTextContent('First skeleton');
      expect(skeletons[1]).toHaveTextContent('Second skeleton');
      expect(skeletons[2]).toHaveTextContent('Third skeleton');
    });

    it('should handle skeletons with different sizes', () => {
      const { container } = render(
        <div>
          <Skeleton className="w-16 h-4">Small</Skeleton>
          <Skeleton className="w-32 h-6">Medium</Skeleton>
          <Skeleton className="w-64 h-8">Large</Skeleton>
        </div>
      );

      const skeletons = getAllByDataSlot(container, 'skeleton');
      expect(skeletons[0]).toHaveClass('w-16', 'h-4');
      expect(skeletons[1]).toHaveClass('w-32', 'h-6');
      expect(skeletons[2]).toHaveClass('w-64', 'h-8');
    });
  });

  describe('Snapshot Testing for UI Consistency', () => {
    it('should match snapshot for default skeleton', () => {
      const { container } = render(<Skeleton />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for skeleton with content', () => {
      const { container } = render(<Skeleton>Loading content</Skeleton>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for skeleton with custom styling', () => {
      const { container } = render(<Skeleton className="w-32 h-8 bg-gray-200" />);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for multiple skeletons', () => {
      const { container } = render(
        <div>
          <Skeleton className="w-16 h-4" />
          <Skeleton className="w-32 h-6" />
          <Skeleton className="w-64 h-8" />
        </div>
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Performance Testing', () => {
    it('should handle rapid re-renders efficiently', () => {
      const { rerender } = render(<Skeleton>Initial skeleton</Skeleton>);

      expect(screen.getByText('Initial skeleton')).toBeInTheDocument();

      // Rerender with different content
      rerender(<Skeleton>Updated skeleton</Skeleton>);

      expect(screen.getByText('Updated skeleton')).toBeInTheDocument();
      expect(screen.queryByText('Initial skeleton')).not.toBeInTheDocument();
    });

    it('should handle className changes efficiently', () => {
      const { rerender, container } = render(<Skeleton className="w-16 h-4">Small</Skeleton>);

      const skeleton = getByDataSlot(container, 'skeleton');
      expect(skeleton).toHaveClass('w-16', 'h-4');

      // Change className
      rerender(<Skeleton className="w-32 h-6">Medium</Skeleton>);

      const updatedSkeleton = getByDataSlot(container, 'skeleton');
      expect(updatedSkeleton).toHaveClass('w-32', 'h-6');
      expect(updatedSkeleton).not.toHaveClass('w-16', 'h-4');
    });

    it('should handle multiple skeleton renders efficiently', () => {
      const skeletons = Array.from({ length: 100 }, (_, i) => `Skeleton ${i + 1}`);

      const startTime = performance.now();

      render(
        <div>
          {skeletons.map((text, index) => (
            <Skeleton key={index} className="w-16 h-4">
              {text}
            </Skeleton>
          ))}
        </div>
      );

      const endTime = performance.now();

      expect(screen.getByText('Skeleton 1')).toBeInTheDocument();
      expect(screen.getByText('Skeleton 100')).toBeInTheDocument();
      expect(endTime - startTime).toBeLessThan(1000); // Should render within 1 second
    });
  });
}); 