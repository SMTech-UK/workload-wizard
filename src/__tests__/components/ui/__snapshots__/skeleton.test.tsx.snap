// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for default skeleton 1`] = `
<div
  class="bg-accent animate-pulse rounded-md"
  data-slot="skeleton"
/>
`;

exports[`Skeleton Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for multiple skeletons 1`] = `
<div>
  <div
    class="bg-accent animate-pulse rounded-md w-16 h-4"
    data-slot="skeleton"
  />
  <div
    class="bg-accent animate-pulse rounded-md w-32 h-6"
    data-slot="skeleton"
  />
  <div
    class="bg-accent animate-pulse rounded-md w-64 h-8"
    data-slot="skeleton"
  />
</div>
`;

exports[`Skeleton Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for skeleton with content 1`] = `
<div
  class="bg-accent animate-pulse rounded-md"
  data-slot="skeleton"
>
  Loading content
</div>
`;

exports[`Skeleton Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for skeleton with custom styling 1`] = `
<div
  class="animate-pulse rounded-md w-32 h-8 bg-gray-200"
  data-slot="skeleton"
/>
`;
